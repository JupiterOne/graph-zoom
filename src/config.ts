import {
  IntegrationExecutionContext,
  IntegrationValidationError,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';
import getOrCreateAPIClient from './getOrCreateAPIClient';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  accountId: {
    type: 'string',
    mask: true,
  },
  clientId: {
    type: 'string',
    mask: true,
  },
  clientSecret: {
    type: 'string',
    mask: true,
  },
  scopes: {
    type: 'string',
  },
};

export interface IntegrationConfig extends IntegrationInstanceConfig {
  accountId: string;
  clientId: string;
  clientSecret: string;
  scopes: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;

  if (
    !config.accountId ||
    !config.clientId ||
    !config.clientSecret ||
    !config.scopes
  ) {
    throw new IntegrationValidationError(
      'Config requires all of { accountId, clientId, clientSecret, scopes }.',
    );
  }

  const apiClient = await getOrCreateAPIClient(config);
  await apiClient.verifyAuthentication();
}
