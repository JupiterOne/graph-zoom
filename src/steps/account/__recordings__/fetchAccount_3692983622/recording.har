{
  "log": {
    "_recordingName": "fetchAccount",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "6.0.5"
    },
    "entries": [
      {
        "_id": "7794742568ef03add90ff15d4508fdbb",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/x-www-form-urlencoded"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "zoom.us"
            }
          ],
          "headersSize": 423,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "queryString": [
            {
              "name": "grant_type",
              "value": "account_credentials"
            },
            {
              "name": "account_id",
              "value": "N0oChUdLTrCZLfk9JXp3Hw"
            }
          ],
          "url": "https://zoom.us/oauth/token?grant_type=account_credentials&account_id=N0oChUdLTrCZLfk9JXp3Hw"
        },
        "response": {
          "bodySize": 756,
          "content": {
            "mimeType": "application/json;charset=UTF-8",
            "size": 756,
            "text": "{\"access_token\":\"[REDACTED]\",\"token_type\":\"bearer\",\"expires_in\":3599,\"scope\":\"user:read:admin account:read:admin group:read:admin role:read:admin\"}"
          },
          "cookies": [
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_aid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_haid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_tmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_htmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "cred",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "httpOnly": true,
              "name": "_zm_page_auth",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "httpOnly": true,
              "name": "_zm_ssid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "2023-07-12T21:54:13.000Z",
              "httpOnly": true,
              "name": "_zm_ctaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "2023-07-12T21:54:13.000Z",
              "httpOnly": true,
              "name": "_zm_chtaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "2025-07-11T19:54:13.000Z",
              "name": "_zm_mtk_guid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "2023-07-12T20:24:13.000Z",
              "httpOnly": true,
              "name": "__cf_bm",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 12 Jul 2023 19:54:13 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json;charset=UTF-8"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "x-zm-trackingid",
              "value": "v=2.0;clid=us06;rid=WEB_bf1303ebfcc3d16e30b6d39f7ac949c6"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-security-policy",
              "value": "upgrade-insecure-requests; default-src https://*.zoom.us https://zoom.us blob: 'self'; img-src https: about: blob: data: 'self'; style-src https: safari-extension: chrome-extension: 'unsafe-inline' data: 'self'; font-src https: safari-extension: chrome-extension: blob: data: 'self'; connect-src * about: blob: data: 'self'; media-src * rtmp: blob: data: 'self'; frame-src https: ms-appx-web: zoommtg: zoomus: wvjbscheme: data: blob: 'self'; object-src 'none'; base-uri 'none';"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "x-zm-zoneid",
              "value": "VA"
            },
            {
              "name": "cache-control",
              "value": "no-store"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            },
            {
              "name": "cf-cache-status",
              "value": "DYNAMIC"
            },
            {
              "name": "report-to",
              "value": "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=efulKEsbtNh1TpszJTwuyGmf4b%2FvIu%2B5MH9OGNYrb%2BBOI9SokZ7jUt%2BYRvDnfvlmakj2CmS%2FPiMwJs7Q6ob8NnYnFkAis954QE9E9H%2FyUDMvXhyXO9A5sdk%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "name": "nel",
              "value": "{\"success_fraction\":0.01,\"report_to\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "name": "server",
              "value": "cloudflare"
            },
            {
              "name": "cf-ray",
              "value": "7e5bd6389ba1e15f-ORD"
            },
            {
              "name": "alt-svc",
              "value": "h3=\":443\"; ma=86400"
            }
          ],
          "headersSize": 2808,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2023-07-12T19:54:12.869Z",
        "time": 247,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 247
        }
      },
      {
        "_id": "32b6c540dbe7339d81f7948939897f76",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.zoom.us"
            }
          ],
          "headersSize": 812,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://api.zoom.us/v2/users/me"
        },
        "response": {
          "bodySize": 830,
          "content": {
            "mimeType": "application/json;charset=UTF-8",
            "size": 830,
            "text": "{\"id\":\"H9Yx0-O8QUCXVIseYAsjaA\",\"first_name\":\"JupiterOne\",\"last_name\":\"Development\",\"display_name\":\"JupiterOne Development\",\"email\":\"development@jupiterone.dev\",\"type\":2,\"role_name\":\"Owner\",\"pmi\":4440405982,\"use_pmi\":false,\"personal_meeting_url\":\"https://us06web.zoom.us/j/4440405982?pwd=c0Jidml0ZS9ZRzA3dEZWSTNHNHlPZz09\",\"timezone\":\"America/New_York\",\"verified\":1,\"dept\":\"\",\"created_at\":\"2021-10-27T00:00:49Z\",\"last_login_time\":\"2023-02-09T16:15:45Z\",\"cms_user_id\":\"\",\"jid\":\"h9yx0-o8qucxviseyasjaa@xmpp.zoom.us\",\"group_ids\":[\"ppUAdwEBRFCdRE0mfLttog\"],\"im_group_ids\":[],\"account_id\":\"N0oChUdLTrCZLfk9JXp3Hw\",\"language\":\"en-US\",\"phone_country\":\"\",\"phone_number\":\"\",\"status\":\"active\",\"job_title\":\"\",\"location\":\"\",\"login_types\":[100],\"role_id\":\"0\",\"account_number\":7011249145,\"cluster\":\"us06\",\"user_created_at\":\"2021-10-27T00:00:49Z\"}"
          },
          "cookies": [
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_aid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_haid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_tmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_htmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "cred",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "_zm_ctaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "_zm_chtaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "2025-07-11T19:54:13.000Z",
              "name": "_zm_mtk_guid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "TS01f92dc5",
              "path": "/",
              "sameSite": "Strict",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "name": "TS018dd1ba",
              "path": "/",
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "2023-07-12T20:24:13.000Z",
              "httpOnly": true,
              "name": "__cf_bm",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 12 Jul 2023 19:54:13 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json;charset=UTF-8"
            },
            {
              "name": "content-length",
              "value": "830"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "x-zm-trackingid",
              "value": "v=2.0;clid=us06;rid=WEB_1cf972353693424c9f3b4a3047c72c9b"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, must-revalidate, no-transform"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "expires",
              "value": "Thu, 01 Jan 1970 00:00:00 GMT"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "vary",
              "value": "Origin, Access-Control-Request-Method, Access-Control-Request-Headers"
            },
            {
              "name": "x-ratelimit-category",
              "value": "Light"
            },
            {
              "name": "x-zm-zoneid",
              "value": "VA"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            },
            {
              "name": "cf-cache-status",
              "value": "DYNAMIC"
            },
            {
              "name": "report-to",
              "value": "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=Biz5gsjU7H9iZJxH%2B050Lh51h3A0L0jRY6PZQPOO4tKl89jd4h%2BLN5yrWla0UZMGE36pCkbRyZOLcqW80TJqLDiaNv%2Fms3lYKPI8%2FONMLyh7AqV0%2Fbcfm1YMdX4g\"}],\"group\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "name": "nel",
              "value": "{\"success_fraction\":0.01,\"report_to\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "name": "server",
              "value": "cloudflare"
            },
            {
              "name": "cf-ray",
              "value": "7e5bd639bfcb0c82-EWR"
            },
            {
              "name": "alt-svc",
              "value": "h3=\":443\"; ma=86400"
            }
          ],
          "headersSize": 2499,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2023-07-12T19:54:13.126Z",
        "time": 208,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 208
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
