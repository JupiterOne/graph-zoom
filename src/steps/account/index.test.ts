import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig } from '../../config';
import { fetchAccount } from '.';
import { integrationConfig } from '../../../test/config';
import { setupZoomRecording } from '../../../test/recording';

describe('#fetchAccount', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupZoomRecording({
      directory: __dirname,
      name: 'fetchAccount',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchAccount(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const account = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('zoom_account'),
    );
    expect(account.length).toBe(1);
    expect(account).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'zoom_account' },
          name: { type: 'string' },
          accountId: { type: 'string' },
        },
      },
    });
  });
});
