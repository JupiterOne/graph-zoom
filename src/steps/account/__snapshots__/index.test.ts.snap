// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchAccount should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "zoom_account:N0oChUdLTrCZLfk9JXp3Hw",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "account_id": "N0oChUdLTrCZLfk9JXp3Hw",
            "account_number": 7011249145,
            "cluster": "us06",
            "cms_user_id": "",
            "created_at": "2021-10-27T00:00:49Z",
            "dept": "",
            "display_name": "JupiterOne Development",
            "email": "development@jupiterone.dev",
            "first_name": "JupiterOne",
            "group_ids": Array [
              "ppUAdwEBRFCdRE0mfLttog",
            ],
            "id": "H9Yx0-O8QUCXVIseYAsjaA",
            "im_group_ids": Array [],
            "jid": "h9yx0-o8qucxviseyasjaa@xmpp.zoom.us",
            "job_title": "",
            "language": "en-US",
            "last_login_time": "2023-02-09T16:15:45Z",
            "last_name": "Development",
            "location": "",
            "login_types": Array [
              100,
            ],
            "personal_meeting_url": "https://us06web.zoom.us/j/4440405982?pwd=c0Jidml0ZS9ZRzA3dEZWSTNHNHlPZz09",
            "phone_country": "",
            "phone_number": "",
            "pmi": 4440405982,
            "role_id": "0",
            "role_name": "Owner",
            "status": "active",
            "timezone": "America/New_York",
            "type": 2,
            "use_pmi": false,
            "user_created_at": "2021-10-27T00:00:49Z",
            "verified": 1,
          },
        },
      ],
      "_type": "zoom_account",
      "accountId": "N0oChUdLTrCZLfk9JXp3Hw",
      "active": true,
      "createdOn": undefined,
      "displayName": "7011249145",
      "id": "H9Yx0-O8QUCXVIseYAsjaA",
      "name": "7011249145",
      "status": "active",
    },
  ],
  "encounteredTypes": Array [
    "zoom_account",
  ],
  "numCollectedEntities": 1,
}
`;
